### PR TITLE
fix(storage): should use epoch with the maximum offset when get

### LINKED ISSUE #14598: bug: spilled data may be skipped on storage get
### Describe the bug

When introducing memtable spill, we have structure kv-entry epoch to be `(epoch_from_barrier, spilled_offset)`. For each `epoch_from_barrier`, we increment the `spilled_offset` per memtable spill to ensure that data spilled within the same `epoch_from_barrier` has different `spilled_offset`

However, in storage `get`, the read epoch provided is purely `epoch_from_barrier` with `spilled_offset == 0` and we use it to [construct a full key](https://github.com/risingwavelabs/risingwave/blob/30821a3f92c78b50e5e6eeb058f162d8f801c4dd/src/storage/src/hummock/store/version.rs#L687), which later will be used in a [SstableIterator::seek](https://github.com/risingwavelabs/risingwave/blob/30821a3f92c78b50e5e6eeb058f162d8f801c4dd/src/storage/src/hummock/mod.rs#L115). This will caused data with `spill_offset > 0` to be skipped.

Example:

Assume we have a SST:
```
key1 (epoch_from_barrier = 1, spilled_offset = 0)
key2 (epoch_from_barrier = 1, spilled_offset = 1)
key3 (epoch_from_barrier = 1, spilled_offset = 1)
```

get key2 with (epoch_from_barrier = 1) should see key2 in the SST but SstableIterator::seek will point to key3 because seek key is `key2 (epoch_from_barrier = 1, spilled_offset = 0)`

### Error message/log

_No response_

### To Reproduce

_No response_

### Expected behavior

_No response_

### How did you deploy RisingWave?

_No response_

### The version of RisingWave

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Fix #14598 

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/hummock/store/version.rs [hunk 1]
```diff
@@ -684,7 +685,11 @@ impl HummockVersionReader {
             Sstable::hash_for_bloom_filter(dist_key.as_ref(), read_options.table_id.table_id())
         });
 
-        let full_key = FullKey::new(read_options.table_id, TableKey(table_key.clone()), epoch);
+        let full_key = FullKey::new_with_gap_epoch(
```
Comment (hzxa21) [inline src/storage/src/hummock/store/version.rs:688]: Can you add some documentations on why we need to use `MAX_SPILL_OFFSET` and add a UT to cover the bug?

**Review by hzxa21 — APPROVED** [review]
Rest LGTM!

File: src/storage/src/hummock/store/version.rs (hunk 1 above)
Comment (hzxa21) [reply src/storage/src/hummock/store/version.rs:690]: How about UT? Are we planning to add it in later PR?

Comment (wcy-fdu) [reply src/storage/src/hummock/store/version.rs:690]: Yes, UT will be added in https://github.com/risingwavelabs/risingwave/pull/14557.

**Pull request merged**